sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/ReadFn.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.cassandra;
+
+import com.datastax.driver.core.Cluster;
+import com.datastax.driver.core.ColumnMetadata;
+import com.datastax.driver.core.PreparedStatement;
+import com.datastax.driver.core.ResultSet;
+import com.datastax.driver.core.Session;
+import com.datastax.driver.core.Token;
+import java.util.Iterator;
+import java.util.stream.Collectors;
+import org.apache.beam.sdk.io.cassandra.CassandraIO.Read;
+import org.apache.beam.sdk.io.cassandra.CassandraIO.ReadAll;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.base.Joiner;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class ReadFn<T> extends DoFn<Read<T>, T> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ReadFn.class);
+
+  private transient Cluster cluster;
+
+  private transient Session session;
+
+  private final ReadAll<T> readAll;
+
+  public ReadFn(ReadAll<T> readAll) {
+    this.readAll = readAll;
```

Comment:
You don't need this ReadAll ref here, Read is enough, ReadAll is the one that delegates to a normal Read.

Reply:
You need this here so we can set up a connection with Setup, this won't work without it.

Reply:
If you want I can spend more time getting performance data, but without caching the connection on setup I saw more than two orders of magnitude slowdown on just the `CassandraIOTest`

Reply:
Caching connections is something we can do in a subsequent PR. Let's get first right the API and then optimize that part. We can cache connections at the class level on the `ReadFn` by creating some size limited pool of connections that we identify by some combination of parameters and we reuse if possible.

Reply:
@iemejia the tests don't even finish running *locally* after ten minutes when running `CassandraIOTest` if we don't cache the connection.  I have already tried doing it without establishing a connection in `setup`, it's not workable.  The CassandraIOTest returns over *200* partitions to return *10* rows.  That means it has to establish over 20 connections and tear them down per row...